iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m

```diff
@@ -496,6 +498,107 @@ - (void)testMessageClickedLaunchesUniqueOutcomeAPIRequest {
+    [NSObjectOverrider runPendingSelectors];
+    // The action should cause an "send tag" API request
+    XCTAssertEqualObjects(OneSignalClientOverrider.lastHTTPRequestType, NSStringFromClass([OSRequestSendTagsToServer class]));
+    XCTAssertEqualObjects(OneSignalClientOverrider.lastHTTPRequest[@"tags"][tagKey], tagKey);
+}
+
+- (void)testMessageClickedLaunchesTagRemoveAPIRequest {
+    let message = [OSInAppMessageTestHelper testMessageJsonWithTriggerPropertyName:OS_DYNAMIC_TRIGGER_KIND_SESSION_TIME withId:@"test_id1" withOperator:OSTriggerOperatorTypeLessThan withValue:@10.0];
+    let registrationResponse = [OSInAppMessageTestHelper testRegistrationJsonWithMessages:@[message]];
+    
+    // the trigger should immediately evaluate to true and should
+    // be shown once the SDK is fully initialized.
+    [OneSignalClientOverrider setMockResponseForRequest:NSStringFromClass([OSRequestRegisterUser class]) withResponse:registrationResponse];
+    
+    [UnitTestCommonMethods initOneSignalAndThreadWait];
+    
+    // the message should now be displayed
+    // simulate a button press (action) on the inapp message
+    let tagKey = @"test1";
+    let action = [OSInAppMessageAction instanceWithJson: @{
+        @"click_type" : @"button",
+        @"id" : @"test_action_id",
+        @"tags" : @{
+                @"removes" : @[tagKey]
+        }
+    }];
+    let testMessage = [OSInAppMessage instanceWithJson:message];
+
+    [OSMessagingController.sharedInstance messageViewDidSelectAction:testMessage withAction:action];
+     // Make sure all 3 sets of tags where send in 1 network call.
+    [NSObjectOverrider runPendingSelectors];
+    [UnitTestCommonMethods runBackgroundThreads];
+    [NSObjectOverrider runPendingSelectors];
+    // The action should cause an "send tag" API request
+    XCTAssertEqualObjects(OneSignalClientOverrider.lastHTTPRequestType, NSStringFromClass([OSRequestSendTagsToServer class]));
+    XCTAssertEqualObjects(OneSignalClientOverrider.lastHTTPRequest[@"tags"][tagKey], @"");
+}
+
+- (void)testMessageClickedLaunchesTagSendAndRemoveAPIRequest {
+    let message = [OSInAppMessageTestHelper testMessageJsonWithTriggerPropertyName:OS_DYNAMIC_TRIGGER_KIND_SESSION_TIME withId:@"test_id1" withOperator:OSTriggerOperatorTypeLessThan withValue:@10.0];
+    let registrationResponse = [OSInAppMessageTestHelper testRegistrationJsonWithMessages:@[message]];
+    
+    // the trigger should immediately evaluate to true and should
+    // be shown once the SDK is fully initialized.
+    [OneSignalClientOverrider setMockResponseForRequest:NSStringFromClass([OSRequestRegisterUser class]) withResponse:registrationResponse];
+    
+    [UnitTestCommonMethods initOneSignalAndThreadWait];
+    
+    // the message should now be displayed
+    // simulate a button press (action) on the inapp message
+    let tagKey = @"test1";
+    let action = [OSInAppMessageAction instanceWithJson: @{
+        @"click_type" : @"button",
```

Comment:
This payload should always have a `close`, `url`, `url_target` even when they are empty strings or nil.

Not a big deal, but could become confusing down the road